draft/real-world-phoenix-lets-deploy-using-kubernetes.md

```diff
@@ -0,0 +1,643 @@
+running. Terraform and Helm. We'll use terraform to setup most of the infrastructure.
+In this way I don't have to remember all the tweaks and install steps I took to
+get everything up and running. I can just automate the whole setup declaratively
+and put the whole thing under version control. 
+This is often referred to these days as an infrastructure-as-code setup.
+I wanted to use terraform for everything, but ran into some trouble setting up a 
+few of the tools needed, so reverted to Helm (the kubernetes package manager) to
+install Traefik and our Gitlab Runner. 
+
+Our Terraform setup will include:
+1. Setting up a DigitalOcean Kubernetes Cluster with 2 worker nodes
+2. Connecting a gitlab project to the kubernetes cluster
+3. Creating a kubernetes_secret to be able to pull images from our private gitlab
+   registry.
+4. Setup a managed database cluster service @ digitalocean
+
+And we'll use Helm to:
+1. Install Traefik as an ingress controller
+2. Create a Gitlab Runner in our cluster 
+
+So let's get crackin'!
+
+### install doctl to talk to your digitalocean account
+
+If we want to automate any of this stuff we'll have to be able to talk to
+digitalocean programmatically, so installing doctl is step one. Please refer to
+this guide to get that setup: [doctl up and running
+guide](https://blog.digitalocean.com/introducing-doctl/)
+
+After you have `doctl` installed, make sure you create a personal access token
+in digitalocean and add that in an environment variable and also initialize your
+account.
+```bash
+export DIGITALOCEAN_TOKEN=[token]
+doctl auth init -t $DIGITALOCEAN_TOKEN
+doctl get account
+```
+
+### Install Terraform
+
+Ofcourse there is a handy install guide for terraform also. Actually just an
+executable, as it is written in go, so they have a handy packaged binary you can use.
+See: [terraform install](https://www.terraform.io/downloads.html) 
+
+### Setup our project
+
+If you want to follow along with this guide, create a folder that will hold your
+terraform configs and create a `main.tf` file that will hold our terraform declarations.
+
+```bash
+mkdir kube-terra
+cd kube-terra
+touch main.tf
+```
+
+### Setup the digitalocean Kubernetes cluster
+
+Terraform works in a declaritive manner, which means that we state what we want
+the world to be like and terraform figures out how to get to that state. In
+terraform you describe an item you want to exist as a resource. A Kubernetes
+cluster is a form of a resource that is provided by one of the many providers
+that exist in terraform. So let's create a digitalocean kubernetes cluster:
+
+Creating a resource is always in the format: `resource [kind] [reference] {}`
+We can use the reference we provide later on to refer back to this resource.
+
+```json
+resource "digitalocean_tag" "kubernetes-cl01" {
+  name = "kubernetes-cl01"
+}
+
+resource "digitalocean_kubernetes_cluster" "cl01" {
+  name    = "cl01"
+  region  = "ams3"
+  version = "1.16.2-do.1"
```

Comment:
This version was no longer available for me. I did have success using the `digitalocean_kubernetes_versions` resource to get the latest 1.16 as described here: https://www.terraform.io/docs/providers/do/d/kubernetes_versions.html